File: mpi-1x-hypack-2013/matrix-vector-matrix-comp-codes/f77-lang-matvect/mv_mult_master_sschd.f
c
c ******************************************************************
c
c		C-DAC Tech Workshop : hyPACK-2013
c                           October 15-18, 2013
c
c  Example 17		: mv_mult_master_sschd.f
c
c  Objective           : Matrix_Vector Multiplication
c                        (Self_scheduling algorithm master program)
c
c  Input               : Process 0 (master) reads files (mdata.inp)
c                        for Matrix and (vdata.inp) for Vector
c
c  Output              : Process 0 prints the result of Matrix_Vector
c                        Multiplication
c
c  Necessary Condition : Number of processors should be greater than
c                        2 and less than or equal to 8
c
c   Created            : August-2013
c
c   E-mail             : hpcfte@cdac.in     
c
c********************************************************************
c
      
      program main

      include 'mpif.h'

      integer MAX_ROWS, MAX_COLS 
      parameter (MAX_ROWS = 1000, MAX_COLS = 1000)
      double precision Matrix(MAX_ROWS,MAX_COLS), Vector(MAX_COLS)
		double precision FinalVector(MAX_COLS)
      double precision Buffer(MAX_COLS), ans

      real*8 Starttime, Endtime, Time
      integer MyRank, Root, Numprocs, ierr, status(MPI_STATUS_SIZE)
      integer i, j, Destination
      integer RowtoSend, Source_tag
      integer NoofRows, NoofCols, VectorSize
      integer Destination_tag, ValidInput

      call MPI_INIT( ierr )
      call MPI_COMM_RANK( MPI_COMM_WORLD, MyRank, ierr )
      call MPI_COMM_SIZE( MPI_COMM_WORLD, Numprocs, ierr )

      Root   = 0
      ValidInput = 1

C    /* .......Read the Matrix From Input file ......*/
      open(unit=12, file = './data/mdata.inp')
      read(12,*) NoofRows, NoofCols
      do i = 1, NoofRows
            read(12,*) (Matrix(i,j),  j=1,NoofCols)
C            write(6,*) (Matrix(i,j), j=1,NoofCols)
      enddo

C    /* Read vector from input file */
      open(unit=13, file = './data/vdata.inp')
      read(13,*) VectorSize
      read(13,*) (Vector(i), i=1,VectorSize)
      do i = 1,VectorSize
C            write(6,*) Vector(i)
      enddo

	close(12)
	close(13)

      if(Numprocs .lt. 2) then
	    print*,"Numprocs less than two .."
	    goto 100
       endif

      if((VectorSize.ne.NoofCols).or.(NoofRows.lt.(Numprocs-1)))then
            ValidInput = 0
      endif

       call MPI_Barrier(MPI_COMM_WORLD, ierr)

       call MPI_Bcast(ValidInput, 1, MPI_INTEGER, Root, 
     $          MPI_COMM_WORLD, ierr)

	 if(ValidInput .eq. 0) then
		 print*,"Invalid input data..... "
		 print*,"NoofCols should be equal to VectorSize"
		 goto 100
         endif

c        send VectorSize and Vector to each other process
         call MPI_BCAST(VectorSize, 1, MPI_INTEGER, Root,
     $        MPI_COMM_WORLD, ierr)

         call MPI_BCAST(Vector, VectorSize, MPI_DOUBLE_PRECISION,
     $        Root, MPI_COMM_WORLD, ierr)

c        send Matrix row to each other process
	      RowToSend = 0
         do 40 i = 1,Numprocs-1
            do 30 j = 1,NoofCols
               Buffer(j) = Matrix(i,j)
30         continue
            call MPI_SEND(Buffer, NoofCols, MPI_DOUBLE_PRECISION, i,
     $           RowtoSend+1, MPI_COMM_WORLD, ierr)
            RowtoSend = RowtoSend+1
40      continue
         
         do 70 i = 1,NoofRows
            call MPI_RECV(ans,1,MPI_DOUBLE_PRECISION,MPI_ANY_SOURCE,
     $           MPI_ANY_TAG,MPI_COMM_WORLD,status,ierr)
            Destination = status(MPI_SOURCE)
            Source_tag = status(MPI_TAG)
            FinalVector(Source_tag) = ans

            if (RowtoSend .lt. NoofRows) then
	      Destination_tag = RowtoSend + 1
               do 50 j = 1,NoofCols
                  Buffer(j) = Matrix(RowtoSend+1,j)
50            continue
               call MPI_SEND(Buffer,NoofCols,MPI_DOUBLE_PRECISION, 
     $              Destination,Destination_tag,MPI_COMM_WORLD,ierr)
               RowtoSend     = RowtoSend+1
            else
            call MPI_SEND(1, 1, MPI_INTEGER, Destination, 0,
     $           MPI_COMM_WORLD, ierr)
            endif
	Endtime = MPI_Wtime()
	Time    = EndTime - Starttime
70      continue

	do i = 1,NoofCols,1
	    write(*, *) 'FinalAns[',i,'] = ',FinalVector(i)
	enddo
	print*,' Time Taken = ',Time

 100   	call MPI_FINALIZE(ierr)

      	stop
      	end

File: mpi-1x-hypack-2013/matrix-vector-matrix-comp-codes/c-lang-matvect/mv_mult_master_sschd.c
/*
********************************************************************

		C-DAC Tech Workshop : hyPACK-2013
                       October 15-18, 2013

   Example 5.5         : mv_mult_master_sschd.c

   Objective           : Matrix_Vector Multiplication
                        (Self_scheduling algorithm master program)

   Input               : Process 0 (master) reads files (mdata.inp) 
                         for Matrix and (vdata.inp) for Vector

   Output              : Process 0 prints the result of Matrix_Vector 
                         Multiplication 

   Necessary Condition : Number of processors should be greater than 
                         2 and less than or equal to 8

   Created             : August-2013

   E-mail              : hpcfte@cdac.in     

****************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>


main(int argc, char** argv) 
{

   
  int	    Numprocs, MyRank;
  int 	    NoofCols, NoofRows, VectorSize; 
  int	    index, irow, icol, iproc, ValidInput=1;
  int	    Root=0, Source, Destination;
  int	    Source_tag, Destination_tag;
  int	    RowtoSend = 0;
  float     **Matrix, *Buffer, Sum;
  float     *Vector, *FinalVector;
  float     *CheckResultVector;
  FILE	    *fp;
  int	    MatrixFileStatus = 1, VectorFileStatus = 1;
  MPI_Status status;     


  /* ........MPI Initialisation .......*/

  MPI_Init(&argc, &argv); 
  MPI_Comm_rank(MPI_COMM_WORLD, &MyRank);
  MPI_Comm_size(MPI_COMM_WORLD, &Numprocs);

  if(Numprocs < 2) {
     printf("Invalid Number of Processors ..... \n");
     printf("Numprocs must be greater than 1 ......\n");
     MPI_Finalize();
     exit(0);
  }  	

  /* .......Read the Input file ......*/
  if ((fp = fopen ("./data/mdata.inp", "r")) == NULL){
       MatrixFileStatus = 0;
  }

  if(MatrixFileStatus != 0) {
     fscanf(fp, "%d %d\n", &NoofRows,&NoofCols);

    /* ...Allocate memory and read Matrix from file .......*/
     Matrix = (float **)malloc(NoofRows*sizeof(float *));
     for(irow=0 ;irow<NoofRows; irow++){
 	 Matrix[irow] = (float *)malloc(NoofCols*sizeof(float));
 	 for(icol=0; icol<NoofCols; icol++) {
    	    fscanf(fp, "%f", &Matrix[irow][icol]);     
       	 }
    }
    fclose(fp);

  }

  /* Read vector from input file */
  if ((fp = fopen ("./data/vdata.inp", "r")) == NULL){
       VectorFileStatus = 0;
  }

  if(VectorFileStatus != 0) {
     fscanf(fp, "%d\n", &VectorSize);

     Vector = (float*)malloc(VectorSize*sizeof(float));
     for(index = 0; index<VectorSize; index++)
       	 fscanf(fp, "%f", &Vector[index]);     

  }

  MPI_Bcast (&MatrixFileStatus, 1, MPI_INT, Root, MPI_COMM_WORLD);
  if(MatrixFileStatus == 0) {
     printf("Can't open input file for Matrix ..... \n");
     MPI_Finalize();
     exit(-1);
  }

  MPI_Bcast (&VectorFileStatus, 1, MPI_INT, Root, MPI_COMM_WORLD);
  if(VectorFileStatus == 0) {
     printf("Can't open input file for Vector ..... \n");
     MPI_Finalize();
     exit(-1);
  }

  if((VectorSize != NoofCols) || (NoofRows < Numprocs-1))
      ValidInput = 0;

  MPI_Bcast(&ValidInput, 1, MPI_INT, Root, MPI_COMM_WORLD);

  if(ValidInput  == 0){
     printf("Invalid input data..... \n");
     printf("NoofCols should be equal to VectorSize\n");
     MPI_Finalize();
     exit(0);
  }

  MPI_Bcast(&VectorSize, 1, MPI_INT, Root, MPI_COMM_WORLD);

  if(MyRank != 0) 
     Vector = (float *)malloc(VectorSize*sizeof(float));

  MPI_Bcast(Vector, VectorSize, MPI_FLOAT, Root, MPI_COMM_WORLD);

  /* .......Allocate memory for output and a row of a matrix......*/ 
  Buffer= (float *)malloc(NoofCols*sizeof(float));
  FinalVector = (float *)malloc(NoofRows*sizeof(float)); 
  for(irow=0; irow< NoofRows; irow++) 
      FinalVector[irow] = 0;

  /* First Send At least one row to each processors */
  for(irow = 1 ; irow < Numprocs ; irow++){
      for(icol=0; icol< NoofCols; icol++){	 
	  Buffer[icol] = Matrix[irow-1][icol];
      }
      MPI_Send( Buffer, NoofCols, MPI_FLOAT, irow, RowtoSend+1, 
		MPI_COMM_WORLD);
      RowtoSend++;
  }

  for(irow = 0 ; irow < NoofRows ; irow++){
      MPI_Recv(&Sum, 1, MPI_FLOAT, MPI_ANY_SOURCE, MPI_ANY_TAG, 
	       MPI_COMM_WORLD, &status);
      Destination = status.MPI_SOURCE;
      Source_tag  = status.MPI_TAG;
      FinalVector[Source_tag] = Sum;

    /*For the  remaining rows(if any) send the data to whichever processor 
        that finishes the earlier computations first */

      if( RowtoSend < NoofRows ){
   	  Destination_tag = RowtoSend+1;
  	  for(icol=0; icol< NoofCols; icol++)
	      Buffer[icol] = Matrix[RowtoSend][icol];

   	  MPI_Send(Buffer, NoofCols, MPI_FLOAT, Destination, 
		   Destination_tag, MPI_COMM_WORLD);
	  RowtoSend++;
      }    
   }

   for(iproc = 1 ; iproc < Numprocs ; iproc++) 
       MPI_Send(Buffer, NoofCols, MPI_FLOAT, iproc, 0, MPI_COMM_WORLD); 
	
   for(irow = 0 ; irow < NoofRows ; irow++) 
       printf("FinalAnswer[%d] =  %f \n", irow, FinalVector[irow]);
		
   free(Vector);
   free(FinalVector);
   free(Buffer);

   MPI_Finalize(); 
}


